bitnami/argo-cd/templates/_helpers.tpl

```diff
@@ -256,6 +256,19 @@ Return the Redis&reg; port
 {{- end -}}
 {{- end -}}
 
+{{/*
+Validate Application Controller config
+*/}}
+{{- define "argocd.validateValues.controller.config" -}}
+{{- if gt (int .Values.controller.replicaCount) 1 }}
+    {{- if and .Values.controller.dynamicClusterDistribution.enabled (not (eq .Values.controller.kind "Deployment")) }}
+Argo CD: When running in HA mode with dynamic cluster distribution enabled, the application controller must be installed as a Deployment.
+    {{- else if and (not .Values.controller.dynamicClusterDistribution.enabled) (not (eq .Values.controller.kind "StatefulSet")) }}
+Argo CD: When running in HA mode, the application controller must be installed as a StatefulSet.
+    {{- end -}}
+{{- end -}}
+{{- end -}}
```

Comment:
I think this validation is not correct or I am understanding the opposite. What should happen if I run this?
```
$ helm install argo-cd bitnami/argo-cd --set server.replicaCount=3 --set controller.replicaCount=3 --set controller.kind=Deployment
```
From your comments I understand that I should face a validation error

Reply:
There is the feature [Dynamic Cluster Distribution](https://argo-cd.readthedocs.io/en/stable/operator-manual/dynamic-cluster-distribution/). If that feature is enabled, the clusters will dynamically be distributed between all of the existing nodes. In this case the `REPLICAS` env variable is ignored.

Reply:
And yes, the command from you above should fail. And it does, right?

Reply:
Hm. Kind-of. The condition is the opposite, since it is printing the message, if the validation fails.

The whole block only renders, when the replica count is > 1. Line 226 is true when dynamic cluster distribution is not enabled and the controller is not deployed as a StatefulSet.

Reply:
Please try it, this is what I see from my side
```console
helm install argo-cd bitnami/argo-cd --set server.replicaCount=3 --set controller.replicaCount=3 --set controller.kind=Deployment
NAME: argo-cd
LAST DEPLOYED: Thu Jul 11 14:13:57 2024
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
CHART NAME: argo-cd
CHART VERSION: 6.6.0
APP VERSION: 2.11.4

** Please be patient while the chart is being deployed **

1. Access your Argo CD installation:
  Execute the following commands:

  kubectl port-forward --namespace default svc/argo-cd-server 8080:80 &
  export URL=http://127.0.0.1:8080/
  echo "Argo CD URL: http://127.0.0.1:8080/"

2. Execute the following commands to obtain the Argo CD credentials:

  echo "Username: \"admin\""
  echo "Password: $(kubectl -n default get secret argocd-secret -o jsonpath="{.data.clearPassword}" | base64 -d)"

WARNING: There are "resources" sections in the chart not set. Using "resourcesPreset" is not recommended for production. For production installations, please set the following values according to your workload needs:
  - controller.resources
  - repoServer.resources
  - server.resources
+info https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
```

Reply:
Yeah. You are right. I think there is something broken with the validations in general... I think the `validateValues` helper misses this block in the end

```
{{- if $message -}}
{{-   printf "\nVALUES VALIDATION:\n%s" $message -}}
{{- end -}}
```

Reply:
At least that is part of some of the other charts, that I checked.

Reply:
And it should probably be `{{-   printf "\nVALUES VALIDATION:\n%s" $message | fail -}}`, so that the deployment fails. This is very in-consequent throughout all the charts though. `fluentd` fails. `flux` does not. `airflow` also does not fail. `logstash` does.

Reply:
I completely agree. Most of our charts fail and print an error message  when the validation is not successful. In this case there was a failure in that piece of code  and I thought the problem was in you code. Sorry about that.

Reply:
Nothing to be sorry about. Did not test this error case tbh. Just thought I can "integrate" in the existing implemtation. But glad we fixed another issue in the chart 🙂